src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,471 @@
+     *     incorrect
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to
+     *     {@code deriveData}
+     */
+    @Override
+    protected byte[] engineDeriveData(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+
+        // inspect KDFParameterSpec object
+        inspectKDFParameterSpec(kdfParameterSpec);
+
+        try {
+            // set up the HMAC instance
+            hmacLen = setupHMAC(hmacAlgName);
+        } catch (NoSuchAlgorithmException nsae) {
+            throw new ProviderException(nsae);
+        }
+
+        if (HKDFTYPE == HKDFTYPES.EXTRACT) {
+            // perform extract
+            try {
+                return hkdfExtract(initialKeyMaterial, (salt == null) ? null : salt.getEncoded());
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Extract could not be initialized with the given key or salt material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXPAND) {
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Expand could not be initialized with the given key material");
+            }
+
+        } else if (HKDFTYPE == HKDFTYPES.EXTRACTEXPAND) {
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(initialKeyMaterial,
+                                                   (salt == null) ? null : salt.getEncoded());
+                this.pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF ExtractExpand could not be initialized with the given key or salt "
+                    + "material");
+            }
+        }
+
+        return null;
+    }
+
+    protected int setupHMAC(String hmacAlgName) throws NoSuchAlgorithmException {
+        hmacObj = Mac.getInstance(hmacAlgName);
+        return hmacObj.getMacLength();
+    }
+
+    protected void inspectKDFParameterSpec(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        // A switch would be nicer, but we may need to backport this before JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable in instanceof either
+        if (kdfParameterSpec instanceof HKDFParameterSpec.Extract) {
```

Comment:
`kdfParameterSpec instanceof HKDFParameterSpec.Extract anExtract`.

Reply:
We should take advantage of newer language features and APIs in the implementation when it simplifies or improves the code. We already are well prepared to handle these cases for other backports. Depending on newer language features or APIs is specific to the API as that could prevent or make it much more difficult to backport.